README.md
# json_transform_model

## 0.0.9
### json 转model


只用一行命令，直接将Json文件转为Dart model类。
```
flutter packages pub run json_transform_model
```

## 安装

```
dependencies:
  flutter:
    sdk: flutter
  ………………
  json_annotation: ^4.0.1

dev_dependencies: 
  json_transform_model: #最新版本
  build_runner: ^2.0.4
  json_serializable: ^4.1.3
```



## 使用

1. 在工程根目录下创建一个名为 "jsons" 的目录;
2. 创建或拷贝json文件到"jsons" 目录中 ;
3. 运行 `pub run json_model` (Dart VM工程)or `flutter packages pub run json_model`(Flutter中) 命令生成Dart model类，生成的文件默认在"lib/models"目录下



## 借鉴json_model生成类文件的功能，在这个基础上把所有的json文件映射到一个文件`JsonConvert`中，对外提供一下api：

```dart
//根据T和给定的字典数据 生成对应的模型
static fromJson<T>(Map<String, dynamic> json){}
//根据T和给定的模型生成 对应的字典
static Map<String, dynamic> toJson<T>() {}
//根据对应的list(json) 批量生成对应的模型数据
static M fromJsonAsT<M>(json){}
```

主要用于在网络请求解析数据的时候提供一个基于泛型的方法，在底层就解析好数据抛给上层来处理。



## 在[`json_model`](https://pub.dev/packages/json_model)的基础上，提供一个JsonConvert文件的一个模板，该模板就是动态生成的，不用修改，

```dart
//模板内容
import '../models/index.dart';

// **************************************************************************
// 动态生成的文件，不要手动去修改
// **************************************************************************

Type typeOf<T>() => T;
class JsonConvert {
  static fromJson<T>(Map<String, dynamic> json) {
    return _getFromJson<T>(typeOf<T>(),T, json);
  }

  static Map<String, dynamic> toJson<T>() {
    return _getToJson<T>(typeOf<T>(), T);
  }

  static _getFromJson<T>(Type type, data, json) {
    switch (type) {
$getFromJson
  }
    return data as T;
  }

  static _getToJson<T>(Type type,data) {
    switch (type) {
$modelToJson
    }
    return data as T;
  }
  //Go back to a single instance by type
  static _fromJsonSingle(String type, json) {
    switch (type) {
$singleJsonToModel
    }
    return null;
  }

  //empty list is returned by type
  static _getListFromType(String type) {
    switch (type) {
$listModel
    }
    return null;
  }

  static M fromJsonAsT<M>(json) {
    String type = M.toString();
    if (json is List && type.contains("List<")) {
      String itemType = type.substring(5, type.length - 1);
      List tempList = _getListFromType(itemType);
      json.forEach((itemJson) {
        tempList
            .add(_fromJsonSingle(type.substring(5, type.length - 1), itemJson));
      });
      return tempList as M;
    } else {
      return _fromJsonSingle(M.toString(), json) as M;
    }
  }
}
  ''';
}
```



## 例子

Json文件: `jsons/user.json`

```
{
  "name":"wendux",
  "father":"$user", //可以通过"$"符号引用其它model类
  "friends":"$[]user", // 可以通过"$[]"来引用数组
  "keywords":"$[]String", // 同上
  "age":20
}
```

生成的Dart model类:

```
import 'package:json_annotation/json_annotation.dart';
part 'user.g.dart';

@JsonSerializable()
class User {
    User(this.name,this.father,this.friends,this.keywords,this.age);
    
    String name;
    User father;
    List<User> friends;
    List<String> keywords;
    num age;
    
    factory User.fromJson(Map<String,dynamic> json) => _$UserFromJson(json);
    Map<String, dynamic> toJson() => _$UserToJson(this);
}
```

### @JsonKey

您也可以使用[json_annotation](https://pub.dev/packages/json_annotation)包中的“@JsonKey”标注特定的字段。

这个功能在特定场景下非常有用，比如Json文件中有一个字段名为"+1"，由于在转成Dart类后，字段名会被当做变量名，但是在Dart中变量名不能包含“+”，我们可以通过“@JsonKey”来映射变量名；

```
{
  "@JsonKey(ignore: true) dynamic":"md",
  "@JsonKey(name: '+1') int": "loved", //将“+1”映射为“loved”
  "name":"wendux",
  "age":20
}
```

生成文件如下:

```
import 'package:json_annotation/json_annotation.dart';
part 'user.g.dart';

@JsonSerializable()
class User {
    User(this.loved,this.name,this.age);
    @JsonKey(name: '+1') int loved;
    String name;
    num age;
    
    factory User.fromJson(Map<String,dynamic> json) => _$UserFromJson(json);
    Map<String, dynamic> toJson() => _$UserToJson(this);
}
```

测试:

```
import 'models/index.dart';

void main() {
  var u = User.fromJson({"name": "Jack", "age": 16, "+1": 20});
  print(u.loved); // 20
}
```

> 关于 `@JsonKey`标注的详细内容请参考[json_annotation](https://pub.dev/packages/json_annotation) 包；

### @Import

另外，提供了一个`@Import `指令，该指令可以在生成的Dart类中导入指定的文件：

```
{
  "@import":"test_dir/profile.dart",
  "@JsonKey(ignore: true) Profile":"profile",
  "name":"wendux",
  "age":20
}
```

生成的Dart类:

```
import 'package:json_annotation/json_annotation.dart';
import 'test_dir/profile.dart';  // 指令生效
part 'user.g.dart';

@JsonSerializable()
class User {
    User(this.profile,this.name,this.age);

    @JsonKey(ignore: true) Profile profile; //file
    String name;
    num age;
    
    factory User.fromJson(Map<String,dynamic> json) => _$UserFromJson(json);
    Map<String, dynamic> toJson() => _$UserToJson(this);
}
```

更完整的示例请移步[这里](https://github.com/flutterchina/json_model/tree/master/example) .

## 命令参数

默认的源json文件目录为根目录下名为 "json" 的目录；可以通过 `src` 参数自定义源json文件目录，例如:

```
pub run json_model src=json_files 
```

默认的生成目录为"lib/models"，同样也可以通过`dist` 参数来自定义输出目录:

```
pub run json_model src=json_files  dist=data # 输出目录为 lib/data
```

> 注意，dist会默认已lib为根目录。

##


CHANGELOG.md
## 0.1.1
### json 转model


只用一行命令，直接将Json文件转为Dart model类。
```
flutter packages pub run json_transform_model
```

## 安装

```
dev_dependencies: 
  json_transform_model: #最新版本
  build_runner: ^2.0.4
  json_serializable: ^5.0.2
```



## 使用

1. 在工程根目录下创建一个名为 "jsons" 的目录;
2. 创建或拷贝json文件到"jsons" 目录中 ;
3. 运行 `pub run json_model` (Dart VM工程)or `flutter packages pub run json_model`(Flutter中) 命令生成Dart model类，生成的文件默认在"lib/models"目录下



## 借鉴json_model生成类文件的功能，在这个基础上把所有的json文件映射到一个文件`JsonConvert`中，对外提供一下api：

```dart
//根据T和给定的字典数据 生成对应的模型
static fromJson<T>(Map<String, dynamic> json){}
//根据T和给定的模型生成 对应的字典
static Map<String, dynamic> toJson<T>() {}
//根据对应的list(json) 批量生成对应的模型数据
static M fromJsonAsT<M>(json){}
```

主要用于在网络请求解析数据的时候提供一个基于泛型的方法，在底层就解析好数据抛给上层来处理。



## 在[`json_model`](https://pub.dev/packages/json_model)的基础上，提供一个JsonConvert文件的一个模板，该模板就是动态生成的，不用修改，

```dart
//模板内容
import '../models/index.dart';

// **************************************************************************
// 动态生成的文件，不要手动去修改
// **************************************************************************

Type typeOf<T>() => T;
class JsonConvert {
  static fromJson<T>(Map<String, dynamic> json) {
    return _getFromJson<T>(typeOf<T>(),T, json);
  }

  static Map<String, dynamic> toJson<T>() {
    return _getToJson<T>(typeOf<T>(), T);
  }

  static _getFromJson<T>(Type type, data, json) {
    switch (type) {
$getFromJson
  }
    return data as T;
  }

  static _getToJson<T>(Type type,data) {
    switch (type) {
$modelToJson
    }
    return data as T;
  }
  //Go back to a single instance by type
  static _fromJsonSingle(String type, json) {
    switch (type) {
$singleJsonToModel
    }
    return null;
  }

  //empty list is returned by type
  static _getListFromType(String type) {
    switch (type) {
$listModel
    }
    return null;
  }

  static M fromJsonAsT<M>(json) {
    String type = M.toString();
    if (json is List && type.contains("List<")) {
      String itemType = type.substring(5, type.length - 1);
      List tempList = _getListFromType(itemType);
      json.forEach((itemJson) {
        tempList
            .add(_fromJsonSingle(type.substring(5, type.length - 1), itemJson));
      });
      return tempList as M;
    } else {
      return _fromJsonSingle(M.toString(), json) as M;
    }
  }
}
  ''';
}
```



## 例子

Json文件: `jsons/user.json`

```
{
  "name":"wendux",
  "father":"$user", //可以通过"$"符号引用其它model类
  "friends":"$[]user", // 可以通过"$[]"来引用数组
  "keywords":"$[]String", // 同上
  "age":20
}
```

生成的Dart model类:

```
import 'package:json_annotation/json_annotation.dart';
part 'user.g.dart';

@JsonSerializable()
class User {
    User(this.name,this.father,this.friends,this.keywords);
    
    String name;
    User father;
    List<User> friends;
    List<String> keywords;
    num age;
    
    factory User.fromJson(Map<String,dynamic> json) => _$UserFromJson(json);
    Map<String, dynamic> toJson() => _$UserToJson(this);
}
```

### @JsonKey

您也可以使用[json_annotation](https://pub.dev/packages/json_annotation)包中的“@JsonKey”标注特定的字段。

这个功能在特定场景下非常有用，比如Json文件中有一个字段名为"+1"，由于在转成Dart类后，字段名会被当做变量名，但是在Dart中变量名不能包含“+”，我们可以通过“@JsonKey”来映射变量名；

```
{
  "@JsonKey(ignore: true) dynamic":"md",
  "@JsonKey(name: '+1') int": "loved", //将“+1”映射为“loved”
  "name":"wendux",
  "age":20
}
```

生成文件如下:

```
import 'package:json_annotation/json_annotation.dart';
part 'user.g.dart';

@JsonSerializable()
class User {
    User(this.loved,this.name,this.age);
    @JsonKey(name: '+1') int loved;
    String name;
    num age;
    
    factory User.fromJson(Map<String,dynamic> json) => _$UserFromJson(json);
    Map<String, dynamic> toJson() => _$UserToJson(this);
}
```

测试:

```
import 'models/index.dart';

void main() {
  var u = User.fromJson({"name": "Jack", "age": 16, "+1": 20});
  print(u.loved); // 20
}
```

> 关于 `@JsonKey`标注的详细内容请参考[json_annotation](https://pub.dev/packages/json_annotation) 包；

### @Import

另外，提供了一个`@Import `指令，该指令可以在生成的Dart类中导入指定的文件：

```
{
  "@import":"test_dir/profile.dart",
  "@JsonKey(ignore: true) Profile":"profile",
  "name":"wendux",
  "age":20
}
```

生成的Dart类:

```
import 'package:json_annotation/json_annotation.dart';
import 'test_dir/profile.dart';  // 指令生效
part 'user.g.dart';

@JsonSerializable()
class User {
    User(this.profile,this.name,this.age);

    @JsonKey(ignore: true) Profile profile; //file
    String name;
    num age;
    
    factory User.fromJson(Map<String,dynamic> json) => _$UserFromJson(json);
    Map<String, dynamic> toJson() => _$UserToJson(this);
}
```

更完整的示例请移步[这里](https://github.com/flutterchina/json_model/tree/master/example) .

## 命令参数

默认的源json文件目录为根目录下名为 "json" 的目录；可以通过 `src` 参数自定义源json文件目录，例如:

```
pub run json_model src=json_files 
```

默认的生成目录为"lib/models"，同样也可以通过`dist` 参数来自定义输出目录:

```
pub run json_model src=json_files  dist=data # 输出目录为 lib/data
```

> 注意，dist会默认已lib为根目录。

##


test/json_transform_model_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:json_transform_model/json_transform_model.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
  });
}


lib/json_convert_genericity.dart
//
//  json_convert
//  flutter_json_to_model
//
//  Created by zhangjiang on 6/4/21 .
//  Copyright © flutter_json_to_model. All rights reserved.

import 'dart:io';
import 'package:path/path.dart' as path;

/// 默认输出目录
const SRC="./jsons"; //JSON 目录
const DIST="lib/models/"; //输出model目录

//定义替换的内容
// const getFromJson = "%{getFromJson}";
// const modelToJson = "%{modelToJson}";
// const singleJsonToModel = "%{singleJsonToModel}";
// const listModel = "%{listModel}";

void walk(String src,String dist) { //遍历JSON目录生成模板
  var srcDire = new Directory(src);
  List<FileSystemEntity> list = srcDire.listSync();
  File file;
  StringBuffer modelToJsonStr = new StringBuffer();
  StringBuffer jsonToModelStr = new StringBuffer();
  StringBuffer signalModelStr = new StringBuffer();
  StringBuffer listModelStr = new StringBuffer();
  StringBuffer indexStr = new StringBuffer();
  list.forEach((f) {
    if (FileSystemEntity.isFileSync(f.path)) {
      file = new File(f.path);
      var paths = path.basename(f.path).split(".");
      String name = paths.first;
      String className = name[0].toUpperCase() + name.substring(1);
      if (paths.last.toLowerCase() != "json" || name.startsWith("_")) return;
      if (name.startsWith("_")) return;
      //设置jsonToModel
      jsonToModelStr.write('    case $className:\r\n');
      jsonToModelStr.write("       return $className.fromJson(json) as T;\r\n");

      //设置modelToJson
      modelToJsonStr.write('    case $className:\r\n');
      modelToJsonStr.write("       return (data as $className).toJson();\r\n");

      //singleModel
      signalModelStr.write("     case '$className':\r\n");
      signalModelStr.write('       return $className.fromJson(json);\r\n');

      //listModel
      listModelStr.write("     case '$className':\r\n");
      listModelStr.write('       return List<$className>.empty();\r\n');

      //添加索引,文件名不用大写
      indexStr.write("export '$name.dart' ; \r\n");
    }
  });
  var content = _getTemplateContent();
  content = replaceContent(content, [jsonToModelStr.toString(),modelToJsonStr.toString(),signalModelStr.toString(),listModelStr.toString()]);
  //将生成的模板输出
  new File("$DIST/JsonConvert.dart").writeAsStringSync(content);
}

String replaceContent(String content,List<Object> params){
  int matchIndex = 0;
  String replace(Match m){
    if (matchIndex < params.length) {
      switch (m[0]){
        case "%s":
          return params[matchIndex++].toString();
      }
    }else{
      throw new Exception("Missing parameter for string format");
    }
    throw new Exception("Invalid format string" + m[0].toString());
  }

  return content.replaceAllMapped('%s', replace);
}


String _getTemplateContent(){
  return '''
  
// **************************************************************************
// 自动生成代码，不要手动修改
  /*
                           _ooOoo_
                          o8888888o
                          88" . "88
                          (| -_- |)
                          O  =  /O
                       ____/`---'____
                     .'  |     |//  `.
                    /  |||  :  |||//  
                   /  _||||| -:- |||||-  
                   |   |   -  /// |   |
                   | _|  ''---/''  |   |
                     .-__  `-`  ___/-. /
                 ___`. .'  /--.--  `. . __
              ."" '<  `.____<|>_/___.'  >'"".
             | | :  `- `.;` _ /`;.`/ - ` : | |
                `-.   _ __ /__ _/   .-` /  /
        ======`-.____`-.________/___.-`____.-'======
                           `=---='
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                 佛祖保佑       永无BUG
* 佛曰:

* 写字楼里写字间，写字间里程序员；

* 程序人员写程序，又拿程序换酒钱。

* 酒醒只在网上坐，酒醉还来网下眠；

* 酒醉酒醒日复日，网上网下年复年。

* 但愿老死电脑间，不愿鞠躬老板前；

* 奔驰宝马贵者趣，公交自行程序员。

* 别人笑我忒疯癫，我笑自己命太贱；

* 不见满街漂亮妹，哪个归得程序员？
// **************************************************************************

import '../models/index.dart';

Type typeOf<T>() => T;
class JsonConvert {
  static fromJson<T>(Map<String, dynamic> json) {
    return _getFromJson<T>(typeOf<T>(),T, json);
  }

  static Map<String, dynamic> toJson<T>() {
    return _getToJson<T>(typeOf<T>(), T);
  }

  static _getFromJson<T>(Type type, data, json) {
    switch (type) {
%s
  }
    return data as T;
  }

  static _getToJson<T>(Type type,data) {
    switch (type) {
%s
    }
    return data as T;
  }
  //Go back to a single instance by type
  static _fromJsonSingle(String type, json) {
    switch (type) {
%s
    }
    return null;
  }

  //empty list is returned by type
  static _getListFromType(String type) {
    switch (type) {
%s
    }
    return null;
  }

  static M fromJsonAsT<M>(json) {
    String type = M.toString();
    if (json is List && type.contains("List<")) {
      String itemType = type.substring(5, type.length - 1);
      List tempList = _getListFromType(itemType);
      json.forEach((itemJson) {
        tempList
            .add(_fromJsonSingle(type.substring(5, type.length - 1), itemJson));
      });
      return tempList as M;
    } else {
      return _fromJsonSingle(M.toString(), json) as M;
    }
  }
}
  ''';
}

void main(){
  walk(SRC,DIST);
}

lib/json_transform_model.dart
//
//  json_model
//  flutter_json_to_model
//
//  Created by zhangjiang on 6/3/21 .
//  Copyright © flutter_json_to_model. All rights reserved.

//定义对应的模型文件路径解析，展示和对应的工具
import 'dart:convert';
import 'dart:io';
import 'json_convert_genericity.dart';
import 'package:path/path.dart' as path;
import 'src/build_runner.dart' as build;
import 'json_convert_genericity.dart' as convert;
import 'package:args/args.dart';

const SRC="./jsons"; //JSON 目录
const Dire="lib/models/";
const tpl='''

// **************************************************************************
// 自动生成代码，不要手动修改
  /*
                           _ooOoo_
                          o8888888o
                          88" . "88
                          (| -_- |)
                          O  =  /O
                       ____/`---'____
                     .'  |     |//  `.
                    /  |||  :  |||//  
                   /  _||||| -:- |||||-  
                   |   |   -  /// |   |
                   | _|  ''---/''  |   |
                     .-__  `-`  ___/-. /
                 ___`. .'  /--.--  `. . __
              ."" '<  `.____<|>_/___.'  >'"".
             | | :  `- `.;` _ /`;.`/ - ` : | |
                `-.   _ __ /__ _/   .-` /  /
        ======`-.____`-.________/___.-`____.-'======
                           `=---='
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                 佛祖保佑       永无BUG
* 佛曰:

* 写字楼里写字间，写字间里程序员；

* 程序人员写程序，又拿程序换酒钱。

* 酒醒只在网上坐，酒醉还来网下眠；

* 酒醉酒醒日复日，网上网下年复年。

* 但愿老死电脑间，不愿鞠躬老板前；

* 奔驰宝马贵者趣，公交自行程序员。

* 别人笑我忒疯癫，我笑自己命太贱；

* 不见满街漂亮妹，哪个归得程序员？
// **************************************************************************

import 'package:json_annotation/json_annotation.dart';
%t

part '%s.g.dart';

@JsonSerializable()
class %s {
    %s(
%s
    );
    
    %s
    
    factory %s.fromJson(Map<String,dynamic> json) => _\$%sFromJson(json);
    Map<String, dynamic> toJson() => _\$%sToJson(this);
}
''';

main(List<String> args){
  run(args);
}

void run(List<String> args){
  String src = SRC;
  String dist = DIST;
  String tag = '\$';
  var parser = new ArgParser();
  parser.addOption('src', defaultsTo: src, callback: (v) => src = v!, help: "Specify the json directory.");
  parser.addOption('dist', defaultsTo: dist, callback: (v) => dist = v!, help: "Specify the dist directory.");
  parser.addOption('tag', defaultsTo: tag, callback: (v) => tag = v!, help: "Specify the tag ");
  parser.parse(args);
  if (walk(src,dist,tag)) {
    //生成jsonConvert文件
    convert.walk(src,dist);
    print("开始生成对应的 .g.dart文件");
    build.run(['build','--delete-conflicting-outputs']);
    print("生成对应的 .g.dart文件完成");
  }
}


//遍历JSON目录生成模板
bool walk(String srcDir, String distDir,String tag) {
  if(srcDir.endsWith("/")) srcDir=srcDir.substring(0, srcDir.length-1);
  if(distDir.endsWith("/")) distDir=distDir.substring(0, distDir.length-1);
  var src = Directory(srcDir);
  var list = src.listSync(recursive: true);
  String indexFile = "";
  if(list.isEmpty) return false;
  if(!Directory(distDir).existsSync()){
    Directory(distDir).createSync(recursive: true);
  }

  print("开始生成对应的JsonSerializable模板");
  File file;
  list.forEach((f) {
    if (FileSystemEntity.isFileSync(f.path)) {
      file = File(f.path);
      var paths = path.basename(f.path).split(".");
      String name = paths.first;
      if(paths.last.toLowerCase()!= "json"||name.startsWith("_")) return ;
      if(name.startsWith("_")) return;
      //下面生成模板
      var map = json.decode(file.readAsStringSync());
      //为了避免重复导入相同的包，我们用Set来保存生成的import语句。
      var set = new Set<String>();
      StringBuffer attrs= new StringBuffer();
      StringBuffer initAttrs = new StringBuffer();
      (map as Map<String, dynamic>).forEach((key, v) {
        if(key.startsWith("_")) return ;
        if(key.startsWith("@")){
          if(key.startsWith("@import")){
            set.add(key.substring(1)+" '$v'");
            return;
          }
          attrs.write(key);
          attrs.write(" ");
          attrs.write(v);
          attrs.writeln(";");
          //设置对应的属性
          initAttrs.write("        this."+key+",\r\n");
        }else {
          attrs.write(getType(v, set, name, tag));
          attrs.write(" ");
          attrs.write(key);
          attrs.writeln(";");

          initAttrs.write("        this."+key+",\r\n");
        }
        attrs.write("    ");
      });
      String className=name[0].toUpperCase()+name.substring(1);
      //替换文本
      var dist = format(tpl,[name,className,className,initAttrs.toString(),attrs.toString(),
        className,className,className]);

      var _import = set.join(";\r\n");
      _import += _import.isEmpty?"":";";
      dist = dist.replaceFirst("%t",_import );
      //将生成的模板输出
      var p = f.path.replaceFirst(srcDir, distDir).replaceFirst(".json", ".dart");
      //写入文件中
      File(p)..createSync(recursive: true)..writeAsStringSync(dist);
      var relative = p.replaceFirst(distDir+path.separator, "");
      indexFile += "export '$relative' ; \n";
    }
  });
  if(indexFile.isNotEmpty) {
    File(path.join(distDir, "index.dart")).writeAsStringSync(indexFile);
  }

  print("JsonSerializable模板生成完毕");
  return indexFile.isNotEmpty;
}


String changeFirstChar(String str, [bool upper=true] ){
  return (upper?str[0].toUpperCase():str[0].toLowerCase()) + str.substring(1);
}

bool isBuiltInType(String type){
  return ['int','num','string','double','map','list'].contains(type);
}

//将JSON类型转为对应的dart类型
String getType(v,Set<String> set,String current, tag){
  current = current.toLowerCase();
  if(v is bool){
    return "bool?";
  }else if(v is num){
    return "num?";
  }else if(v is Map){
    return "Map<String,dynamic>?";
  }else if(v is List){
    return "List?";
  }else if(v is String){ //处理特殊标志
    if(v.startsWith("$tag[]")){
      var type = changeFirstChar(v.substring(3),false);
      if(type.toLowerCase() != current&&!isBuiltInType(type)) {
        set.add('import "$type.dart"');
      }
      return "List<${changeFirstChar(type)}>?";

    }else if(v.startsWith(tag)){
      var fileName = changeFirstChar(v.substring(1),false);
      if(fileName.toLowerCase()!= current) {
        set.add('import "$fileName.dart"');
      }
      return changeFirstChar(fileName);
    }else if(v.startsWith("@")){
      return v;
    }
    return "String?";
  }else{
    return "String?";
  }
}

// String getDefaultValueByType(String type,String className){
//   if (type == 'bool') {
//     return 'false';
//   }else if (type == 'num') {
//     return '0';
//   }else if (type == 'String') {
//     return "\"\"";
//   }else if (type == "List") {
//     return 'List.empty()';
//   }else if (type.contains("Map<")) {
//     return 'Map<String,dynamic>()';
//   }else if (type.contains("List<")) {
//     return 'List.empty()';
//   }
//   return "";
// }

//替换模板占位符
String format(String fmt, List<Object> params) {
  int matchIndex = 0;
  String replace(Match m) {
    if (matchIndex < params.length) {
      switch (m[0]) {
        case "%s":
          return params[matchIndex++].toString();
      }
    } else {
      throw new Exception("Missing parameter for string format");
    }
    throw new Exception("Invalid format string: " + m[0].toString());
  }
  return fmt.replaceAllMapped("%s", replace);
}
